### PR TITLE
(Alternative to 1140): Record layers as incomplete before trying to create them

### PR DESCRIPTION
**Pull request by mtrmac**
This implements the suggestion in #1139, along with **EDIT partially** fixing #1147 ; as an alternative to #1140 .

Conceptually I like this much better than #1140, but it’s quite a bit more invasive, and potentially risky by invoking the layer removal code in previously-unreachable situations.

Marking as draft because it’s completely untested so far.

### CODE REVIEW

**Comment by rhatdan** [comment]
Should we open a PR against this in Podman to get its test suite to run against it?

File: layers.go [hunk 1]
```diff
@@ -732,105 +731,102 @@ func (r *layerStore) Put(id string, parentLayer *Layer, names []string, mountLab
...
+		r.byname[name] = layer
+	}
+	for flag, value := range flags {
+		layer.Flags[flag] = value
+	}
+	layer.Flags[incompleteFlag] = true
+
+	succeeded := false
+	cleanupFailureContext := ""
+	defer func() {
```
Comment (mtrmac) [inline layers.go:761]: Note that using `defer` means that we would try to delete the WIP layer even on a panic. That’s probably correct for some banal `nil` dereference, but might make things worse if the in-memory state were significantly corrupt.

**Comment by mtrmac** [comment]
> Marking as draft because it’s completely untested so far.

Tested now, at least for success and observing the intermediate states. Note that any ambition to being crash-resilient requires a solution to #1136 .

**Comment by mtrmac** [comment]
> Should we open a PR against this in Podman to get its test suite to run against it?

Good idea, https://github.com/containers/podman/pull/13316 . (And similarly for the other PR.)

**Comment by mtrmac** [comment]
Marking as ready to review — please consider this _only_ in tandem with #1140.

**Comment by mtrmac** [comment]
@nalind @giuseppe PTAL. This might be very wrong, I’d just like to make sure _something_ is done about these failure cases.

**Comment by nalind** [comment]
This looks like it has a decent chance of avoiding the types of cases that #1140 would have to clean up on behalf of binaries using a version of the storage library that doesn't include this change, so it looks like it's worth pursuing.

**Comment by rhatdan** [comment]
Merging both PRs with agreement from @nalind and @mtrmac

**Pull request merged**